### PR TITLE
Use codecs

### LINKED ISSUE #3: Add support to codecs
This plugin doesn't support codec as the formating output and uses on the old `#format_message` and `json_message` methods. We need to upgrade the code, add test coverage for it and update the documentation.

### PR DESCRIPTION
**Pull request by andrewvc**
WIP. I still need to add tests here. Fixes #3 . It's now in a structure that is amenable to testing however.

### CODE REVIEW

**Comment by andrewvc** [comment]
@ph would you be able to take a look at this? This includes a full test suite. It can be manually tested with a logstash config like:

```
input {
  generator {
    lines => [
      "Hello 100",
      "test 24.8ms"
    ]
    count => 1
  }
}

filter {
  grok {
    match => [ "message", "test (N/A|%{BASE10NUM:duration:float}ms)" ]
  }
}

output {
  sns {
    arn => "arn"
    access_key_id => "accesskey"
    secret_access_key => "secret"
    codec => rubydebug
    region => "us-east-1"
  }
}
```

or 

```
input {
  generator {
    lines => [
      "Hello 100",
      "test 24.8ms"
    ]
    count => 1
  }
}

filter {
  mutate {
    add_field => { "sns_message" => "Hey dude!" }
  }
}


output {
  stdout { codec => rubydebug}

  sns {
    arn => "key"
    access_key_id => "key"
    secret_access_key => "secret"
    codec => rubydebug
    region => "us-east-1"
  }
}
```

**Comment by ph** [comment]
@andrewvc I have edited your comment to remove your credentials, please invalid them!

File: lib/logstash/outputs/sns.rb [hunk 1]
```diff
@@ -70,54 +63,45 @@ def register
...
+      @sns.topics[@publish_boot_message_arn].publish("Logstash successfully booted", :subject => "Logstash booted")
+    end
   end
 
-  def self.json_message(event)
-    json      = event.to_json
-    json_size = json.bytesize
+  private
+  def send_sns_message(arn, subject, message)
+    raise ArgumentError, "An SNS ARN is required." unless arn
```
Comment (ph) [inline lib/logstash/outputs/sns.rb]: I would move this validation up in the `#register` so logstash would refuse to start before processing anything, doing the check inline with the `#receive` methods would make logstash lose events.

File: lib/logstash/outputs/sns.rb [hunk 2]
```diff
@@ -25,7 +25,7 @@
 #  `MAX_MESSAGE_SIZE_IN_BYTES`.
 #
 class LogStash::Outputs::Sns < LogStash::Outputs::Base
-  include LogStash::PluginMixins::AwsConfig
+  include LogStash::PluginMixins::AwsConfig::V2
 
   MAX_SUBJECT_SIZE_IN_CHARACTERS  = 100
   MAX_MESSAGE_SIZE_IN_BYTES       = 32768
```
Comment (ph) [inline lib/logstash/outputs/sns.rb:30]: Since we have changed the truncate logic it is a good time to update the documentation for the `format` option and mention when the truncate occur and we do it.

File: lib/logstash/outputs/sns.rb [hunk 3]
```diff
@@ -70,54 +63,45 @@ def register
...
-
-    message = Array(event["sns_message"]).first
-    subject = Array(event["sns_subject"]).first || event["host"]
-
-    # Ensure message doesn't exceed the maximum size.
-    if message
-      # TODO: Utilize `byteslice` in JRuby 1.7: http://jira.codehaus.org/browse/JRUBY-5547
-      message = message.slice(0, MAX_MESSAGE_SIZE_IN_BYTES)
+    if (sns_msg = Array(event["sns_message"]).first)
+      send_sns_message(event_arn(event), event_subject(event), sns_msg)
```
Comment (ph) [inline lib/logstash/outputs/sns.rb]: Instead of creating 'adhoc' methods for everything we could create only one that accept the key and return the first element? What do you think? If this is use case is often used we could move this helper to the `Event` class. (first time I see it.)

I see the clojure inspiration here, deconstruction everywhere! ;)

File: lib/logstash/outputs/sns.rb [hunk 4]
```diff
@@ -70,54 +63,45 @@ def register
...
-
-    # Publish the message.
-    @sns.topics[arn].publish(message, :subject => subject.slice(0, MAX_SUBJECT_SIZE_IN_CHARACTERS))
+  private
+  def publish_boot_message_arn
+    # Try to publish a "Logstash booted" message to the ARN provided to
+    # cause an error ASAP if the credentials are bad.
+    if @publish_boot_message_arn
+      @sns.topics[@publish_boot_message_arn].publish("Logstash successfully booted", :subject => "Logstash booted")
+    end
```
Comment (ph) [inline lib/logstash/outputs/sns.rb:81]: This method could use the `send_sns_message` for sending the event?

File: spec/outputs/sns_spec.rb [hunk 5]
```diff
@@ -2,17 +2,110 @@
 require "logstash/devutils/rspec/spec_helper"
 require 'logstash/outputs/sns'
 require 'logstash/event'
+require "logstash/plugin_mixins/aws_config"
+
+require "aws-sdk" # TODO: Why is this not automatically brought in by the aws_config plugin?
```
Comment (ph) [inline spec/outputs/sns_spec.rb:7]: You sure you need that?  The `register` method include the `aws-sdk-resource`,
But I see your point, I am not a fan of lazy require since they are lazy per default, because plugins are only loaded when they are defined in a running configuration.

File: spec/outputs/sns_spec.rb [hunk 6]
```diff
@@ -2,17 +2,110 @@
...
     end
 
-    it 'should allow to output to sns with a list of tags' do
-      event = LogStash::Event.new({"tags" => ["elasticsearch", "logstash", "kibana"] })
-      expect(LogStash::Outputs::Sns.format_message(event)).to match(/\nTags:\selasticsearch,\slogstash,\skibana\n/m)
+    describe "with an explicit message" do
+      let(:expected_message) { sns_message }
+      let(:event) { LogStash::Event.new("sns" => arn, "sns_subject" => sns_subject, "sns_message" => sns_message) }
+      include_examples("publishing correctly")
     end
```
Comment (ph) [inline spec/outputs/sns_spec.rb:54]: good use of `shared_examples` +1

File: spec/outputs/sns_spec.rb [hunk 7]
```diff
@@ -2,17 +2,110 @@
...
+
+    describe "without an explicit message" do
+      # Testing codecs sucks. It'd be nice if codecs had to implement some sort of encode_sync method
+      let(:expected_message) {
```
Comment (ph) [inline spec/outputs/sns_spec.rb:58]: Agree with this, I don't recall the history why we only have an async method. We don't have anything async.
This would make the code easier and cleaner IMHO.

Comment (ph) [inline spec/outputs/sns_spec.rb:58]: We could add a new one and keep the async for backward compatibly.

File: spec/outputs/sns_spec.rb [hunk 8]
```diff
@@ -2,17 +2,110 @@
...
+    it "should send a well formed message through to SNS" do
+      expect(mock_client).to receive(:publish).with(good_publish_args)
+      subject.send(:send_sns_message, arn, sns_subject, sns_message)
+    end
+
+    it "should attempt to publish a boot message" do
+      expect(subject).to have_received(:publish_boot_message_arn).once
+    end
+
+    # TODO: Write a unicode aware version of this...
```
Comment (ph) [inline spec/outputs/sns_spec.rb]: I prefer creating issues instead of `TODO`, they get hidden in the code :(

File: spec/outputs/sns_spec.rb [hunk 9]
```diff
@@ -2,17 +2,110 @@
...
+
+    it "should truncate long subjects before sending" do
+      max_size = LogStash::Outputs::Sns::MAX_SUBJECT_SIZE_IN_CHARACTERS
+      long_subject = "A" * (max_size + 1)
+      expect(mock_client).to receive(:publish) {|args|
+                               expect(args[:subject].bytesize).to eql(max_size)
+                             }
+
+      subject.send(:send_sns_message, arn, long_subject, sns_message)
+      end
```
Comment (ph) [inline spec/outputs/sns_spec.rb]: A bit too many spaces before the `end`

File: spec/outputs/sns_spec.rb [hunk 10]
```diff
@@ -2,17 +2,110 @@
...
+
+    it "should attempt to publish a boot message" do
+      expect(subject).to have_received(:publish_boot_message_arn).once
+    end
+
+    # TODO: Write a unicode aware version of this...
+    it "should truncate long messages before sending" do
+      max_size = LogStash::Outputs::Sns::MAX_MESSAGE_SIZE_IN_BYTES
+      long_message = "A" * (max_size + 1)
+      expect(mock_client).to receive(:publish) {|args|
```
Comment (ph) [inline spec/outputs/sns_spec.rb:126]: I think we should extract the `long_message` in a let statement.

File: spec/outputs/sns_spec.rb [hunk 11]
```diff
@@ -2,17 +2,110 @@
...
+                               expect(args[:message].bytesize).to eql(max_size)
+                             }
+
+      subject.send(:send_sns_message, arn, sns_subject, long_message)
+    end
+
+    it "should truncate long subjects before sending" do
+      max_size = LogStash::Outputs::Sns::MAX_SUBJECT_SIZE_IN_CHARACTERS
+      long_subject = "A" * (max_size + 1)
```
Comment (ph) [inline spec/outputs/sns_spec.rb]: I think we should extract the `long_subject` in a let statement.

File: CHANGELOG.md [hunk 12]
```diff
@@ -0,0 +1,3 @@
+# 0.1.5
+  * If no `subject` are specified fallback to the %{host} key (https://github.com/logstash-plugins/logstash-output-sns/pull/2)
+  * Migrate the SNS Api to use the AWS-SDK v2
```
Comment (ph) [inline CHANGELOG.md:6]: Update this with your changes!

**Comment by ph** [comment]
Good work @andrewvc ! I added a few comments and I didn't not run the test since some of the changes will probably break them :(

File: lib/logstash/outputs/sns.rb (hunk 3 above)
Comment (andrewvc) [inline lib/logstash/outputs/sns.rb]: @ph I'm kind of confused by this myself. I didn't realize that any field in logstash could be an array. Is this code correct in this? Should it behave this way at all?

File: lib/logstash/outputs/sns.rb (hunk 4 above)
Comment (andrewvc) [inline lib/logstash/outputs/sns.rb:81]: Yep! Good catch

File: lib/logstash/outputs/sns.rb (hunk 1 above)
Comment (andrewvc) [inline lib/logstash/outputs/sns.rb]: I agree with the idea here, but I think that would violate the contract of the plugin API. The original author had it publish a boot message to essentially test the connection, so this check will be run anyway.

Additionally, this plugin can direct the output to an ARN specified by a message, so we need to do check on each send anyway.

One could, in fact, setup this plugin with the boot message disabled (it is configurable) and explicit ARNs sent with each message (maybe via add_field), in which case a check in register would break such a configuration.

File: lib/logstash/outputs/sns.rb (hunk 2 above)
Comment (andrewvc) [inline lib/logstash/outputs/sns.rb:30]: Good point! Should we actually just get rid of the format option since now `@codec` works?

File: spec/outputs/sns_spec.rb (hunk 5 above)
Comment (andrewvc) [inline spec/outputs/sns_spec.rb:7]: Ah I needed that there for tests to run, since some of my tests don't call `#register`

File: spec/outputs/sns_spec.rb (hunk 6 above)
Comment (andrewvc) [inline spec/outputs/sns_spec.rb:54]: ty :)

File: spec/outputs/sns_spec.rb (hunk 7 above)
Comment (andrewvc) [inline spec/outputs/sns_spec.rb:58]: Awesome, I'll add a ticket to Logstash core for this.

Comment (andrewvc) [inline spec/outputs/sns_spec.rb:58]: Ticket added: https://github.com/elastic/logstash/issues/3486

File: spec/outputs/sns_spec.rb (hunk 8 above)
Comment (andrewvc) [inline spec/outputs/sns_spec.rb]: There is a PR I made with the todo. I'll excise the todo in my next push.

File: lib/logstash/outputs/sns.rb (hunk 3 above)
Comment (ph) [inline lib/logstash/outputs/sns.rb]: Any fields can be an `Array` or a `Hash`, I think what we should do here is to allow the user to set as an option which fields they want for the `sns_message` or `sns_subject` and we assume that the field is a string. 

By doing `.first` we already do a decision for the user, if they want to use the first value of the array they can still do it:

``` ruby
irb(main):006:0> require "logstash/event"
=> false
irb(main):007:0> e = LogStash::Event.new({"awesome" => ["one", "two"], "notawesome" => "wifi-on-train"})
=> #<LogStash::Event:0x4aee6e0a @metadata_accessors=#<LogStash::Util::Accessors:0x36e0a11d @store={}, @lut={}>, @cancelled=false, @data={"awesome"=>["one", "two"], "notawesome"=>"wifi-on-train", "@version"=>"1", "@timestamp"=>"2015-06-22T18:30:28.766Z"}, @metadata={}, @accessors=#<LogStash::Util::Accessors:0x2ed46cd5 @store={"awesome"=>["one", "two"], "notawesome"=>"wifi-on-train", "@version"=>"1", "@timestamp"=>"2015-06-22T18:30:28.766Z"}, @lut={}>>
irb(main):008:0> e["[awesome][0]"]
=> "one"
irb(main):009:0> e["[notawesome][0]"]
=> nil
irb(main):010:0>
```

This change isn't backward compatible, but I think this is an uncommon situation..
Also since we uses v2 api and we now uses codec which are major changes I believe the plugin should be released as a new major version.

Maybe other people can comment on this?

File: spec/outputs/sns_spec.rb (hunk 5 above)
Comment (ph) [inline spec/outputs/sns_spec.rb:7]: Well they should be calling register so the plugin is in the correct state no?

File: lib/logstash/outputs/sns.rb (hunk 1 above)
Comment (ph) [inline lib/logstash/outputs/sns.rb]: Well if the `publish_boot_message_arn` use this method it will be run :)

Comment (ph) [inline lib/logstash/outputs/sns.rb]: OK I agree since messages can override the routing information, make sense to have it here.

File: lib/logstash/outputs/sns.rb (hunk 3 above)
Comment (andrewvc) [inline lib/logstash/outputs/sns.rb]: I agree that we should just break this functionality and make the decision explicitly. I think a good default behavior would be to log a JSON representation of `Array|Hash` values.

Comment (andrewvc) [inline lib/logstash/outputs/sns.rb]: Actually, my last comment makes no sense for the message field. For that we should pass it through the codec, and whatever comes out goes in. For the subject however, we should either fail if the message is not a string, or save the subject as JSON. 

I'm leaning toward the JSON option.

Comment (suyograo) [inline lib/logstash/outputs/sns.rb]: From the API spec here: http://docs.aws.amazon.com/sns/latest/APIReference/API_Publish.html

`subject` needs to be a string? My vote is to save the message to JSON (instead of dropping events) if it is not a string and mark it with `invalid_subject_type` tag so users can review such events and correct it

Comment (andrewvc) [inline lib/logstash/outputs/sns.rb]: @suyograo makes sense. @ph I think that gets what you were getting at as well. I'll go ahead and execute along these lines.

Comment (ph) [inline lib/logstash/outputs/sns.rb]: +1

File: spec/outputs/sns_spec.rb (hunk 5 above)
Comment (andrewvc) [inline spec/outputs/sns_spec.rb:7]: Well, this let block for instance needs it since setup a mock on the Aws classes themselves:

```
  let(:instance) {
    allow(Aws::SNS::Client).to receive(:new).and_return(mock_client)
    inst = LogStash::Outputs::Sns.new
    allow(inst).to receive(:publish_boot_message_arn).and_return(nil)
    inst.register
    inst
  }
```

If you remove the require the tests will fail unfortunately.

**Comment by andrewvc** [comment]
Thanks for all the great feedback @ph  and @suyograo I've made the requested changes and improved the docs to boot. I also covered a corner case that had been overlooked (no subject AND no host). More specs are now in place as well.

File: lib/logstash/outputs/sns.rb (hunk 2 above)
Comment (andrewvc) [inline lib/logstash/outputs/sns.rb:30]: I did wind up updating the docs. I also removed the format option.

File: CHANGELOG.md (hunk 12 above)
Comment (ph) [inline CHANGELOG.md:6]: Update the changelog mentioning the codec.

Comment (andrewvc) [inline CHANGELOG.md:6]: Done!

**Comment by ph** [comment]
@andrewvc @suyograo I would be inclined to use version 2.0 for this plugin instead of marking it as 1.0 because we have decided that plugins shipped are already marked as 1.0. 

And the 1.0 is using v1, so 2.0.0 would make sense to me for a v2. API.

Other than this minor comment code LGTM, Tests are an awesome improvement +1.
I manually tested this plugin also.

We could probably try https://github.com/logstash-plugins/logstash-output-sns/issues/4 in another PR.
